File: matlab/+rtls/AnchorConfig.m
classdef AnchorConfig < handle
    %ANCHORCONFIG  Anchor coordinates + room geometry.
    %
    % THIS IS THE ONE PLACE YOU ADD ANCHORS. The firmware only ships ranges, so
    % adding/moving an anchor is purely a host-side config change here (or in the
    % JSON file) - no reflash needed.
    %
    %   cfg = rtls.AnchorConfig();                 % built-in 3-anchor example
    %   cfg = rtls.AnchorConfig.fromJson('config/anchors.json');
    %
    %   [A, found] = cfg.coordsFor([1 2 3]);       % map ids -> coordinates

    properties
        dim    = 2          % 2 or 3
        ids    = []         % 1xK anchor short addresses (decimal)
        coords = zeros(0,3) % Kx3 [x y z] in metres (z ignored when dim==2)
        bounds = [0 5 0 5 0 3]  % [xmin xmax ymin ymax zmin zmax] for plotting
    end

    methods
        function obj = AnchorConfig(ids, coords, dim, bounds)
            if nargin == 0
                % Built-in example: 3 anchors around a ~4x4 m room, dim=2.
                obj.dim    = 2;
                obj.ids    = [1, 2, 3];
                obj.coords = [0.00, 0.00, 1.50;
                              4.00, 0.00, 1.50;
                              2.00, 4.00, 1.50];
                obj.bounds = [-0.5 4.5 -0.5 4.5 0 3];
                return;
            end
            obj.ids    = ids(:)';
            obj.coords = coords;
            if nargin >= 3 && ~isempty(dim);    obj.dim = dim; end
            if nargin >= 4 && ~isempty(bounds); obj.bounds = bounds; end
        end

        function [A, found] = coordsFor(obj, queryIds)
            %COORDSFOR  Rows of coordinates for the requested ids.
            % A is numel(queryIds) x dim; found is a logical mask of known ids.
            q = queryIds(:)';
            A = zeros(numel(q), obj.dim);
            found = false(1, numel(q));
            for i = 1:numel(q)
                idx = find(obj.ids == q(i), 1);
                if ~isempty(idx)
                    A(i, :) = obj.coords(idx, 1:obj.dim);
                    found(i) = true;
                end
            end
        end

        function c = centroid(obj)
            c = mean(obj.coords(:, 1:obj.dim), 1);
        end
    end

    methods (Static)
        function obj = fromJson(path)
            txt = fileread(path);
            j = jsondecode(txt);
            ids = zeros(1, numel(j.anchors));
            coords = zeros(numel(j.anchors), 3);
            for i = 1:numel(j.anchors)
                a = j.anchors(i);
                ids(i) = a.id;
                coords(i, :) = [a.x, a.y, getfield_default(a, 'z', 0)];
            end
            dim = getfield_default(j, 'dim', 2);
            bounds = getfield_default(j, 'bounds', [0 5 0 5 0 3]);
            obj = rtls.AnchorConfig(ids, coords, dim, bounds(:)');
        end
    end
end

function v = getfield_default(s, f, d)
    if isfield(s, f) && ~isempty(s.(f)); v = s.(f); else; v = d; end
end

File: python/rtls/anchor_config.py
"""
Anchor coordinates + room geometry.

Reads config/anchors.json (same format as the MATLAB version).
"""

import json
import numpy as np
from pathlib import Path
from typing import Tuple, List


class AnchorConfig:
    def __init__(self, ids=None, coords=None, dim=2, bounds=None):
        if ids is None:
            # Built-in default: 4-anchor square, 4 m side, centred at origin.
            # A1=(-2,-2), A2=(2,-2), A3=(2,2), A4=(-2,2) — all on z=0 ground plane.
            self.dim    = 2
            self.ids    = [1, 2, 3, 4]
            self.coords = np.array([
                [-2.0, -2.0, 0.0],
                [ 2.0, -2.0, 0.0],
                [ 2.0,  2.0, 0.0],
                [-2.0,  2.0, 0.0],
            ])
            self.bounds = [-2.5, 2.5, -2.5, 2.5, 0.0, 3.0]
        else:
            self.dim    = dim
            self.ids    = list(ids)
            self.coords = np.array(coords, dtype=float)
            self.bounds = bounds if bounds else [0, 5, 0, 5, 0, 3]

    @classmethod
    def from_json(cls, path: str) -> 'AnchorConfig':
        with open(path) as f:
            j = json.load(f)
        anchors = j['anchors']
        ids    = [a['id'] for a in anchors]
        coords = [[a['x'], a['y'], a.get('z', 0.0)] for a in anchors]
        dim    = j.get('dim', 2)
        bounds = j.get('bounds', [0, 5, 0, 5, 0, 3])
        return cls(ids, coords, dim, bounds)

    def coords_for(self, query_ids: List[int]) -> Tuple[np.ndarray, List[bool]]:
        """Map anchor ids → coordinates. Returns (Nxdim array, found mask)."""
        A     = np.zeros((len(query_ids), self.dim))
        found = [False] * len(query_ids)
        for i, qid in enumerate(query_ids):
            if qid in self.ids:
                idx = self.ids.index(qid)
                A[i] = self.coords[idx, :self.dim]
                found[i] = True
        return A, found

    def centroid(self) -> np.ndarray:
        return np.mean(self.coords[:, :self.dim], axis=0)
